plugin.py

```diff
@@ -0,0 +1,164 @@
+import sublime
+
+import os
+import re
+import base64
+from urllib.parse import urlparse
+import weakref
+
+from LSP.plugin.core.protocol import Request, Location
+from LSP.plugin.core.sessions import Session
+from LSP.plugin.core.typing import Any, Callable, Dict, List, Mapping, Optional, Tuple, Union
+from LSP.plugin.core.registry import LspTextCommand
+from LSP.plugin import register_plugin, unregister_plugin
+from lsp_utils import NpmClientHandler
+
+
+class LspGrammarlyCommand(LspTextCommand):
+    session_name = 'grammarly'
+    msg_prefix = "LSP-Grammarly: "
+    weaksession = None  # type: Optional[weakref.ref[Session]]
+    cmd = None  # type: Optional[str]
+
+    def error_message(self, msg: str) -> None:
+        sublime.error_message(self.msg_prefix + msg)
+
+    def message_dialog(self, msg: str) -> None:
+        sublime.message_dialog(self.msg_prefix + msg)
+
+    def send_request(self, params: Any) -> None:
+        def run_async() -> None:
+            if not self.cmd:
+                return
+            session = self.session_by_name()
+            if not session:
+                return
+            self.weaksession = weakref.ref(session)
+            request = Request(self.cmd, params, None, progress=True)
+            session.send_request(request, self._on_success_async, self._on_error_async)
+        sublime.set_timeout_async(run_async)
+
+    def _on_success_async(self, response: Any) -> None:
+        raise NotImplementedError()
+
+    def _on_error_async(self, error: Any) -> None:
+        pass
+
+
+class LspGrammarlyExecuteLoginCommand(LspGrammarlyCommand):
+    cmd = "$/getOAuthUrl"
+    cmd_callback = "$/handleOAuthCallbackUri"
+
+    internal_redirect_uri = "vscode://znck.grammarly/auth/callback"
+    external_redirect_uri = internal_redirect_uri
+    redirect_uri = internal_redirect_uri
+
+    def run(self, edit: sublime.Edit) -> None:
+        self.send_request(self.redirect_uri)
```

Comment:
Done a small refactor here to remove duplication (3 identical blocks for sending requests). Now the base class provides `send_request` and sub-classes can override `_on_success_async` and `_on_error_async` to handle request results.

Let me know if you don't like it then we can revert.

Reply:
of course I like it :) thank you for the effort and the patience!